source/geod24/concurrency.d

```diff
@@ -1,1059 +1,982 @@
+    scheduled kernel threads, fibers (cooperative user-space threads), or some
+    other concept with similar behavior.
+
+    he type of concurrency used when logical threads are created is determined
+    by the Scheduler selected at initialization time.  The default behavior is
+    currently to create a new kernel thread per call to spawn, but other
+    schedulers are available that multiplex fibers across the main thread or
+    use some combination of the two approaches.
+
+    Note:
+    Copied (almost verbatim) from Phobos at commit 3bfccf4f1 (2019-11-27)
+    Changes are this notice, and the module rename, from `std.concurrency`
+    to `geod24.concurrency`.
+    Removed Tid, spawn
+    Added Channel
+
+    Copyright: Copyright Sean Kelly 2009 - 2014.
+    License:   <a href="http://www.boost.org/LICENSE_1_0.txt">Boost License 1.0</a>.
+    Authors:   Sean Kelly, Alex Rønne Petersen, Martin Nowak
+    Source:    $(PHOBOSSRC std/concurrency.d)
+
+    Copyright Sean Kelly 2009 - 2014.
+    Distributed under the Boost Software License, Version 1.0.
+    (See accompanying file LICENSE_1_0.txt or copy at
+    http://www.boost.org/LICENSE_1_0.txt)
+
+*******************************************************************************/
+
 module geod24.concurrency;
 
-public import std.variant;
+import std.container;
+import std.range;
 
-import core.atomic;
 import core.sync.condition;
 import core.sync.mutex;
 import core.thread;
-import std.range.primitives;
-import std.range.interfaces : InputRange;
-import std.traits;
 
-///
-@system unittest
-{
-    __gshared string received;
-    static void spawnedFunc(Tid self, Tid ownerTid)
-    {
-        import std.conv : text;
-        // Receive a message from the owner thread.
-        self.receive((int i){
-            received = text("Received the number ", i);
-
-            // Send a message back to the owner thread
-            // indicating success.
-            send(ownerTid, true);
-        });
-    }
 
-    // Start spawnedFunc in a new thread.
-    auto childTid = spawn(&spawnedFunc, thisTid);
-    auto self = thisTid();
+/*******************************************************************************
 
-    // Send the number 42 to this new thread.
-    send(childTid, 42);
+    Encapsulates all implementation-level data needed for scheduling.
 
-    // Receive the result code.
-    auto wasSuccessful = self.receiveOnly!(bool);
-    assert(wasSuccessful);
-    assert(received == "Received the number 42");
-}
+    When defining a Scheduler, an instance of this struct must be associated
+    with each logical thread.  It contains all implementation-level information
+    needed by the internal API.
+
+*******************************************************************************/
 
-private
+public struct ThreadInfo
 {
-    bool hasLocalAliasing(Types...)()
-    {
-        import std.typecons : Rebindable;
+    ///
+    public InfoObject[string] objectValues;
```

Comment:
I see you used a similar pattern in every instances of this:

- A `set` and a `get` `@property` function, using a name;
- The `get` function uses the name and then casts the received value;

I don't see why an AA with interface is useful here, because you don't really have a use for iterating the objects.
Instead of doing:
```D
InfoObject[string] objectValues;
objectValue["scheduler"] = new FiberScheduler;
objectValue["waitingmanager"] = new WaitingManager;
```

Why not just do:
```
struct Something
{
    WaitingManager waiting_manager;
    Scheduler scheduler;
}
Something o;
o.scheduler = ...;
```

In this case, since `objectValues` already lives in a `struct`, I don't think you need the extra struct, you can just store them in `ThreadInfo`, can't you ?

Reply:
Transceiver uses ‘Request’ and ‘Response’.
And in the last review, we did not want to incorporate ‘Request’, ‘Response’ and ‘Status’ into the concurrency.d.
It also decided not to use std.variant. So we used the Associated Array.
By the way, it will all be solved using  Minivariant.

Reply:
Minivariant won't solve the problem at core. There's three common techniques we can use here. I'm going to go over them so we share the same vocabulary.

I) "Functional abstraction": This is something you do when you want to abstract an operation, and that operation has a set of inputs (the parameters) and output (the return value and side effects) that is fixed. This is something `interface`, and OOP is general, is great at. Have a `class`/`interface` with virtual method(s) and just call them.
Note that I've put "Functional abstraction" between quotes, because I'm not actually sure what the *right* name is in this context.

II) [Structural typing](https://en.wikipedia.org/wiki/Structural_type_system) (OR [duck typing](https://en.wikipedia.org/wiki/Duck_typing) depending on your approach, who you ask).
First, let's talk about nominal type systems. D's type system is nominal. When you use the name `Transceiver`, D will look that name up according to a set of rules. If two names match with the same priority, you got a conflict. So you can have things like:
```D
module foo;

struct Bar {}

class FooBar {
    static struct Bar {}

    void method () {
         Bar b; // Refers to `foo.FooBar.Bar`
         .Bar b1; // Refers to `foo.Bar` because we use the "module scope" operator (dot)
         FooBar.Bar b2; // Refers to `foo.FooBar.Bar`

         Bar oops = getBar(); // This won't compile, because you are trying to assign a `foo.Bar` to a `foo.FooBar.Bar`
    }
}

Bar getBar () { return Bar.init; }
```

However, not all type systems are nominal. Some are "structural", which means they depend only on the structure, not on the actual name you give them. Templates are an example of such an approach:
```D
template GetBar (T)
{
    alias GetBar = T.Bar;
}

struct ExternalBar {}
struct Foo { alias Bar = ExternalBar; }
struct Other { struct Bar {} }
struct Func { void Bar() {} }
```
In this example, you can call `GetBar` with any of `Foo`, `Other`, `Func` and it'll compile, because `GetBar` only cares about the structure of `T` (because `T` is not actually the name of the type). So while D's type system is nominal, we can do some structural things by using templates (among other things).

III) Type erasure (/repainting: don't look at the wikipedia article, it's all wrong).
This is simply bypassing the type system to do whatever we want. It's mandatory in some languages (e.g. C), and for some construct, but should be seldom used. For example:
```D
void main ()
{
    int[4] foo;
    func(foo.ptr);
}

void func (void* ptr)
{
    // Here there are two assumption: `ptr` points to an array of `int`, and the array has size 4
    auto slice = (cast(int*)ptr)[0 .. 4];
    // Uses slice...
}
```

----------------------

Now all of those techniques have their ups and downs.

Functional abstraction is more rigorous but require more work (because you have to specify *everything* in the interface). Sometimes it also blows up because you have to split your abstractions into smaller abstractions, when you try to add a primitive that not all types support.
The structural approach is more forgiving, as it doesn't ask you to define everything beforehand, and you can add some special-casing here and there without the code turning into a gigantic mess. From my experience, it is the most common one in D. It has downsides, for example it can quickly get out of control and leak implementation details in the interface (this is why we have `RemoteAPI!(API).spawn!(Implementation)`: If `Implementation` was part of `RemoteAPI`'s template parameters, it would expose too much internal details).

Now the type erasure / painting approach is sometimes used, and *sometimes* useful, but should be used *as little as possible*. It is always the easiest approach to take (because it just makes the compiler shut up), but since it removes all checks being done, it is also the easiest to misuse. So every time you see a `cast`, you should ask yourself: "Is there a way to do this that does not involve a `cast` ?". In 99% of cases, there is, and it's also a *better* way.

TL;DR: Never do type erasure, go for structural (templates) by default, and use functional abstraction when convenient.

Reply:
Okay, so that was the long explanation. The reason I posted it is simple: `InfoObject` is an overly limited functional abstraction (it only abstracts the fact that this object has a finalizer that can be called explicitly), so in order for things to work you have to couple it with type erasure, which is not a good thing.
Instead, you want to *expose* a structural abstraction (template) from `concurrency`.
Note that `Variant` did type erasure too, so that's why some refactor in the client code are necessary.